src/search-and-replace-dialog.jsx

```diff
@@ -0,0 +1,286 @@
+
+        this.setState({ searchResultCount });
+        this.props.onChange({ searchIndex });
+    },
+
+    getSearchResultCount(searchString) {
+        if (searchString === "") {
+            return 0;
+        }
+
+        var count = 0;
+
+        if (this.props.question) {
+            count += Util.countOccurrences(this.props.question.content, searchString);
+        }
+
+        if (this.props.hints) {
+            this.props.hints.forEach(hint => {
+                count += Util.countOccurrences(hint.content, searchString);
+            });
+        }
+
+        if (this.props.particle) {
+            var particle = this.props.particle;
+            if (Array.isArray(particle)) {
+                particle.forEach(section => {
+                    if (section.content) {
+                        count += Util.countOccurrences(section.content, searchString);
+                    }
+                });
+            } else {
+                if (particle.content) {
+                    count += Util.countOccurrences(particle.content, searchString);
+                }
+            }
+        }
+
+        return count;
+    },
+
+    handleNextSearchResult() {
+        var searchIndex = this.props.searchIndex;
+        searchIndex ++;
+        searchIndex = searchIndex % this.state.searchResultCount;
+
+        this.props.onChange({ searchIndex });
+    },
+
+    handlePreviousSearchResult() {
+        var searchIndex = this.props.searchIndex;
+        searchIndex --;
+        if (searchIndex < 0) {
+            searchIndex = this.state.searchResultCount - 1;
+        }
+
+        this.props.onChange({ searchIndex });
+    },
+
+    handleReplaceAll() {
+        var replaceString = this.state.replaceString;
+        var { searchString, question, hints, particle } = this.props;
+
+        var regex = new RegExp(Util.escapeRegExp(searchString), "g");
+
+        var replaceFunc = function(obj) {
+            if (obj) {
+                if (Array.isArray(obj)) {
+                    obj.forEach(replaceFunc);
+                } else {
+                    // excludes occurrences within widget references
+                    var indices = Util.getIndicesOf(obj.content, searchString);
+                    obj.content = obj.content.replace(regex, (match, offset) => {
+                        // make sure this match isn't inside a widget reference
+                        if (indices.indexOf(offset) !== -1) {
+                            return replaceString
+                        } else {
+                            return match;
+                        }
+                    });
+                }
+            }
+        };
+
+        replaceFunc(question);
+        replaceFunc(hints);
+        replaceFunc(particle);
+
+        // handle the case where the replaceString contains one or more copies
+        // of searchString
+        // TODO(kevinb7) this is quite right
+        var searchResultCount = this.getSearchResultCount(searchString);
```

Comment:
I don't think this is going to work once we fix the `replaceFunc` mutating props above, since this method relies on props. We'll probably have to pass those fields to this function (either individually or as a collective object).